// @cypress/code-coverage/dist/lib/common-utils.js
"use strict";
/// <reference types="node" />
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultNycOptions = void 0;
exports.combineNycOptions = combineNycOptions;
exports.fileCoveragePlaceholder = fileCoveragePlaceholder;
exports.removePlaceholders = removePlaceholders;
function stringToArray(prop, obj) {
    if (typeof obj[prop] === 'string') {
        obj[prop] = [obj[prop]];
    }
}
function combineNycOptions(...options) {
    // last option wins
    const nycOptions = Object.assign({}, ...options);
    // normalize string and [string] props
    stringToArray('reporter', nycOptions);
    stringToArray('extension', nycOptions);
    stringToArray('exclude', nycOptions);
    return nycOptions;
}
exports.defaultNycOptions = {
    'report-dir': './coverage',
    reporter: ['lcov', 'clover', 'json', 'json-summary'],
    extension: ['.js', '.cjs', '.mjs', '.ts', '.tsx', '.jsx'],
    excludeAfterRemap: false
};
/**
 * Returns an object with placeholder properties for files we
 * do not have coverage yet. The result can go into the coverage object
 *
 * @param fullPath Filename
 */
function fileCoveragePlaceholder(fullPath) {
    return {
        path: fullPath,
        statementMap: {},
        fnMap: {},
        branchMap: {},
        s: {},
        f: {},
        b: {}
    };
}
function isPlaceholder(entry) {
    // when the file has been instrumented, its entry has "hash" property
    return !('hash' in entry);
}
/**
 * Given a coverage object with potential placeholder entries
 * inserted instead of covered files, removes them. Modifies the object in place
 */
function removePlaceholders(coverage) {
    Object.keys(coverage).forEach((key) => {
        if (isPlaceholder(coverage[key])) {
            delete coverage[key];
        }
    });
}
