// @cypress/code-coverage/dist/lib/support-utils.js
"use strict";
/// <reference types="cypress" />
/// <reference lib="dom" />
Object.defineProperty(exports, "__esModule", { value: true });
exports.filterFilesFromCoverage = filterFilesFromCoverage;
exports.fixSourcePaths = fixSourcePaths;
exports.getSendCoverageBatchSize = getSendCoverageBatchSize;
/** excludes files that shouldn't be in code coverage report */
function filterFilesFromCoverage(totalCoverage, config = Cypress.config, expose = Cypress.expose, spec = Cypress.spec) {
    const withoutSpecs = filterSpecsFromCoverage(totalCoverage, config, expose, spec);
    const appCoverageOnly = filterSupportFilesFromCoverage(withoutSpecs, config);
    return appCoverageOnly;
}
/**
 * remove coverage for the spec files themselves,
 * only keep "external" application source file coverage
 */
function filterSpecsFromCoverage(totalCoverage, config = Cypress.config, expose = Cypress.expose, spec = Cypress.spec) {
    const testFilePatterns = getCypressExcludePatterns(config, expose);
    const isTestFile = (_, filePath) => {
        const workingDir = spec.absolute.replace(spec.relative, '');
        const filename = filePath.replace(workingDir, '');
        const matchedPattern = testFilePatterns.some((specPattern) => Cypress.minimatch(filename, specPattern, { debug: false }));
        const matchedEndOfPath = testFilePatterns.some((specPattern) => filename.endsWith(specPattern));
        return matchedPattern || matchedEndOfPath;
    };
    const coverage = Cypress._.omitBy(totalCoverage, isTestFile);
    return coverage;
}
/**
 * Reads Cypress config and exclude patterns and combines them into one array
 */
function getCypressExcludePatterns(config, expose) {
    const testFilePattern = config('specPattern');
    const codeCoverageConfig = expose().codeCoverage;
    const excludePattern = codeCoverageConfig?.exclude;
    let testFilePatterns = [];
    if (Array.isArray(testFilePattern)) {
        // Filter out any undefined values and ensure all are strings
        for (const p of testFilePattern) {
            if (p !== undefined && typeof p === 'string') {
                testFilePatterns.push(p);
            }
        }
    }
    else if (testFilePattern && typeof testFilePattern === 'string') {
        testFilePatterns = [testFilePattern];
    }
    // combine test files pattern and exclude patterns into single exclude pattern
    if (Array.isArray(excludePattern)) {
        // Filter out any undefined values and ensure all are strings
        for (const p of excludePattern) {
            if (p !== undefined && typeof p === 'string') {
                testFilePatterns.push(p);
            }
        }
    }
    else if (typeof excludePattern === 'string') {
        testFilePatterns.push(excludePattern);
    }
    return testFilePatterns;
}
/**
 * Removes support file from the coverage object.
 * Support files are now handled via the exclude patterns in getCypressExcludePatterns.
 */
function filterSupportFilesFromCoverage(totalCoverage, config = Cypress.config) {
    const supportFile = config('supportFile');
    if (!supportFile) {
        return totalCoverage;
    }
    const isSupportFile = (filename) => filename === supportFile;
    const coverage = Cypress._.omitBy(totalCoverage, (_fileCoverage, filename) => isSupportFile(filename));
    return coverage;
}
/**
 * Replace source-map's path by the corresponding absolute file path
 * (coverage report wouldn't work with source-map path being relative
 * or containing Webpack loaders and query parameters)
 */
function fixSourcePaths(coverage) {
    Object.values(coverage).forEach((file) => {
        const { path: absolutePath, inputSourceMap } = file;
        if (!inputSourceMap || !absolutePath)
            return;
        const fileNameMatch = /([^\/\\]+)$/.exec(absolutePath);
        if (!fileNameMatch)
            return;
        const fileName = fileNameMatch[1];
        if (inputSourceMap.sourceRoot)
            inputSourceMap.sourceRoot = '';
        inputSourceMap.sources = inputSourceMap.sources.map((source) => source.includes(fileName) ? absolutePath : source);
    });
}
/**
 * Validates and returns the configured batch size for
 * sending coverage to the backend
 */
function getSendCoverageBatchSize() {
    const batchSize = Cypress.expose('sendCoverageBatchSize');
    const parsedBatchSize = parseInt(String(batchSize), 10);
    const isValid = !isNaN(parsedBatchSize) && parsedBatchSize > 0;
    return isValid ? parsedBatchSize : null;
}
