// @cypress/code-coverage/dist/lib/task.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
/// <reference types="node" />
const istanbul_lib_coverage_1 = require("istanbul-lib-coverage");
const path_1 = require("path");
const fs_1 = require("fs");
const execa_1 = __importDefault(require("execa"));
const task_utils_1 = require("./task-utils");
const support_utils_1 = require("./support-utils");
const debug_1 = __importDefault(require("debug"));
const log = (0, debug_1.default)('code-coverage');
// these are standard folder and file names used by NYC tools
const processWorkingDirectory = process.cwd();
// there might be custom "nyc" options in the user package.json
// see https://github.com/istanbuljs/nyc#configuring-nyc
// potentially there might be "nyc" options in other configuration files
// it allows, but for now ignore those options
const pkgFilename = (0, path_1.join)(processWorkingDirectory, 'package.json');
const pkg = (0, fs_1.existsSync)(pkgFilename)
    ? JSON.parse((0, fs_1.readFileSync)(pkgFilename, 'utf8'))
    : {};
const scripts = pkg.scripts || {};
const DEFAULT_CUSTOM_COVERAGE_SCRIPT_NAME = 'coverage:report';
const customNycReportScript = scripts[DEFAULT_CUSTOM_COVERAGE_SCRIPT_NAME];
const nycReportOptions = (function getNycOption() {
    // https://github.com/istanbuljs/nyc#common-configuration-options
    const nycReportOptions = (0, task_utils_1.readNycOptions)(processWorkingDirectory);
    if (nycReportOptions.exclude && !Array.isArray(nycReportOptions.exclude)) {
        console.error('NYC options: %o', nycReportOptions);
        throw new Error('Expected "exclude" to be an array');
    }
    if (nycReportOptions['temp-dir']) {
        nycReportOptions['temp-dir'] = (0, path_1.resolve)(nycReportOptions['temp-dir']);
    }
    else {
        nycReportOptions['temp-dir'] = (0, path_1.join)(processWorkingDirectory, '.nyc_output');
    }
    nycReportOptions.tempDir = nycReportOptions['temp-dir'];
    if (nycReportOptions['report-dir']) {
        nycReportOptions['report-dir'] = (0, path_1.resolve)(nycReportOptions['report-dir']);
    }
    // seems nyc API really is using camel cased version
    nycReportOptions.reportDir = nycReportOptions['report-dir'];
    return nycReportOptions;
})();
const nycFilename = (0, path_1.join)(nycReportOptions['temp-dir'], 'out.json');
let coverageMap = (() => {
    const previousCoverage = (0, fs_1.existsSync)(nycFilename)
        ? JSON.parse((0, fs_1.readFileSync)(nycFilename, 'utf8'))
        : {};
    return (0, istanbul_lib_coverage_1.createCoverageMap)(previousCoverage);
})();
function saveCoverage(coverage) {
    if (!(0, fs_1.existsSync)(nycReportOptions.tempDir)) {
        (0, fs_1.mkdirSync)(nycReportOptions.tempDir, { recursive: true });
        log('created folder %s for output coverage', nycReportOptions.tempDir);
    }
    (0, fs_1.writeFileSync)(nycFilename, JSON.stringify(coverage, null, 2));
}
function maybePrintFinalCoverageFiles(folder) {
    const jsonReportFilename = (0, path_1.join)(folder, 'coverage-final.json');
    if (!(0, fs_1.existsSync)(jsonReportFilename)) {
        log('Did not find final coverage file %s', jsonReportFilename);
        return;
    }
    log('Final coverage in %s', jsonReportFilename);
    const finalCoverage = JSON.parse((0, fs_1.readFileSync)(jsonReportFilename, 'utf8'));
    const finalCoverageKeys = Object.keys(finalCoverage);
    log('There are %d key(s) in %s', finalCoverageKeys.length, jsonReportFilename);
    finalCoverageKeys.forEach((key) => {
        const s = finalCoverage[key].s || {};
        const statements = Object.keys(s);
        const totalStatements = statements.length;
        let coveredStatements = 0;
        statements.forEach((statementKey) => {
            if (s[statementKey]) {
                coveredStatements += 1;
            }
        });
        const hasStatements = totalStatements > 0;
        const allCovered = coveredStatements === totalStatements;
        const coverageStatus = hasStatements ? (allCovered ? '✅' : '⚠️') : '❓';
        log('%s %s statements covered %d/%d', coverageStatus, key, coveredStatements, totalStatements);
    });
}
const tasks = {
    /**
     * Clears accumulated code coverage information.
     *
     * Interactive mode with "cypress open"
     *    - running a single spec or "Run all specs" needs to reset coverage
     * Headless mode with "cypress run"
     *    - runs EACH spec separately, so we cannot reset the coverage
     *      or we will lose the coverage from previous specs.
     */
    resetCoverage({ isInteractive }) {
        if (isInteractive) {
            log('reset code coverage in interactive mode');
            coverageMap = (0, istanbul_lib_coverage_1.createCoverageMap)({});
            saveCoverage(coverageMap);
        }
        /*
            Else:
              in headless mode, assume the coverage file was deleted
              before the `cypress run` command was called
              example: rm -rf .nyc_output || true
          */
        return null;
    },
    /**
     * Combines coverage information from single test
     * with previously collected coverage.
     *
     * @param sentCoverage Stringified coverage object sent by the test runner
     * @returns Nothing is returned from this task
     */
    combineCoverage(sentCoverage) {
        const coverage = JSON.parse(sentCoverage);
        log('parsed sent coverage');
        (0, support_utils_1.fixSourcePaths)(coverage);
        coverageMap.merge(coverage);
        return null;
    },
    /**
     * Saves coverage information as a JSON file and calls
     * NPM script to generate HTML report
     */
    coverageReport() {
        saveCoverage(coverageMap);
        if (!(0, fs_1.existsSync)(nycFilename)) {
            console.warn('Cannot find coverage file %s', nycFilename);
            console.warn('Skipping coverage report');
            return null;
        }
        (0, task_utils_1.showNycInfo)(nycFilename);
        const allSourceFilesMissing = (0, task_utils_1.checkAllPathsNotFound)(nycFilename);
        if (allSourceFilesMissing) {
            (0, task_utils_1.tryFindingLocalFiles)(nycFilename);
        }
        (0, task_utils_1.resolveRelativePaths)(nycFilename);
        if (customNycReportScript) {
            log('saving coverage report using script "%s" from package.json, command: %s', DEFAULT_CUSTOM_COVERAGE_SCRIPT_NAME, customNycReportScript);
            log('current working directory is %s', process.cwd());
            return (0, execa_1.default)('npm', ['run', DEFAULT_CUSTOM_COVERAGE_SCRIPT_NAME], {
                stdio: 'inherit'
            }).then(() => null);
        }
        if (nycReportOptions.all) {
            log('nyc needs to report on all included files');
            (0, task_utils_1.includeAllFiles)(nycFilename, nycReportOptions);
        }
        log('calling NYC reporter with options %o', nycReportOptions);
        log('current working directory is %s', process.cwd());
        const NYC = require('nyc');
        const nyc = new NYC(nycReportOptions);
        const returnReportFolder = () => {
            const reportFolder = nycReportOptions['report-dir'];
            log('after reporting, returning the report folder name %s', reportFolder);
            maybePrintFinalCoverageFiles(reportFolder);
            return reportFolder;
        };
        return nyc.report().then(returnReportFolder);
    }
};
module.exports = function registerCodeCoverageTasks(on, config) {
    on('task', tasks);
    // set a variable to let the hooks running in the browser
    // know that they can send coverage commands
    config.expose.codeCoverageTasksRegistered = true;
    return config;
};
