// @cypress/code-coverage/dist/lib/task-utils.js
"use strict";
// helper functions to use from "task.js" plugins code
// that need access to the file system
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.readNycOptions = readNycOptions;
exports.checkAllPathsNotFound = checkAllPathsNotFound;
exports.showNycInfo = showNycInfo;
exports.resolveRelativePaths = resolveRelativePaths;
exports.findCommonRoot = findCommonRoot;
exports.tryFindingLocalFiles = tryFindingLocalFiles;
exports.includeAllFiles = includeAllFiles;
/// <reference types="node" />
const fs_1 = require("fs");
const path_1 = require("path");
const debug_1 = __importDefault(require("debug"));
const chalk_1 = __importDefault(require("chalk"));
const tinyglobby_1 = require("tinyglobby");
const js_yaml_1 = __importDefault(require("js-yaml"));
const common_utils_1 = require("./common-utils");
const log = (0, debug_1.default)('code-coverage');
function readNycOptions(workingDirectory) {
    const pkgFilename = (0, path_1.join)(workingDirectory, 'package.json');
    const pkg = (0, fs_1.existsSync)(pkgFilename)
        ? JSON.parse((0, fs_1.readFileSync)(pkgFilename, 'utf8'))
        : {};
    const pkgNycOptions = pkg.nyc || {};
    const nycrcFilename = (0, path_1.join)(workingDirectory, '.nycrc');
    const nycrc = (0, fs_1.existsSync)(nycrcFilename)
        ? JSON.parse((0, fs_1.readFileSync)(nycrcFilename, 'utf8'))
        : {};
    const nycrcJsonFilename = (0, path_1.join)(workingDirectory, '.nycrc.json');
    const nycrcJson = (0, fs_1.existsSync)(nycrcJsonFilename)
        ? JSON.parse((0, fs_1.readFileSync)(nycrcJsonFilename, 'utf8'))
        : {};
    const nycrcYamlFilename = (0, path_1.join)(workingDirectory, '.nycrc.yaml');
    let nycrcYaml = {};
    if ((0, fs_1.existsSync)(nycrcYamlFilename)) {
        try {
            nycrcYaml = js_yaml_1.default.load((0, fs_1.readFileSync)(nycrcYamlFilename, 'utf8'));
        }
        catch (error) {
            throw new Error(`Failed to load .nycrc.yaml: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
    const nycrcYmlFilename = (0, path_1.join)(workingDirectory, '.nycrc.yml');
    let nycrcYml = {};
    if ((0, fs_1.existsSync)(nycrcYmlFilename)) {
        try {
            nycrcYml = js_yaml_1.default.load((0, fs_1.readFileSync)(nycrcYmlFilename, 'utf8'));
        }
        catch (error) {
            throw new Error(`Failed to load .nycrc.yml: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
    const nycConfigFilename = (0, path_1.join)(workingDirectory, 'nyc.config.js');
    let nycConfig = {};
    if ((0, fs_1.existsSync)(nycConfigFilename)) {
        try {
            nycConfig = require(nycConfigFilename);
        }
        catch (error) {
            throw new Error(`Failed to load nyc.config.js: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
    const nycConfigCommonJsFilename = (0, path_1.join)(workingDirectory, 'nyc.config.cjs');
    let nycConfigCommonJs = {};
    if ((0, fs_1.existsSync)(nycConfigCommonJsFilename)) {
        try {
            nycConfigCommonJs = require(nycConfigCommonJsFilename);
        }
        catch (error) {
            throw new Error(`Failed to load nyc.config.cjs: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
    const nycOptions = (0, common_utils_1.combineNycOptions)(common_utils_1.defaultNycOptions, nycrc, nycrcJson, nycrcYaml, nycrcYml, nycConfig, nycConfigCommonJs, pkgNycOptions);
    log('combined NYC options %o', nycOptions);
    return nycOptions;
}
function checkAllPathsNotFound(nycFilename) {
    const nycCoverage = JSON.parse((0, fs_1.readFileSync)(nycFilename, 'utf8'));
    const coverageKeys = Object.keys(nycCoverage);
    if (!coverageKeys.length) {
        log('⚠️ file %s has no coverage information', nycFilename);
        return;
    }
    const allFilesAreMissing = coverageKeys.every((key) => {
        const coverage = nycCoverage[key];
        return !(0, fs_1.existsSync)(coverage.path);
    });
    log('in file %s all files are not found? %o', nycFilename, allFilesAreMissing);
    return allFilesAreMissing;
}
/**
 * A small debug utility to inspect paths saved in NYC output JSON file
 */
function showNycInfo(nycFilename) {
    const nycCoverage = JSON.parse((0, fs_1.readFileSync)(nycFilename, 'utf8'));
    const coverageKeys = Object.keys(nycCoverage);
    if (!coverageKeys.length) {
        console.error('⚠️ file %s has no coverage information', chalk_1.default.yellow(nycFilename));
        console.error('Did you forget to instrument your web application? Read %s', chalk_1.default.blue('https://github.com/cypress-io/code-coverage#instrument-your-application'));
        return;
    }
    log('NYC file %s has %d key(s)', nycFilename, coverageKeys.length);
    const maxPrintKeys = 3;
    const showKeys = coverageKeys.slice(0, maxPrintKeys);
    showKeys.forEach((key, k) => {
        const coverage = nycCoverage[key];
        // printing a few found keys and file paths from the coverage file
        // will make debugging any problems much much easier
        if (k < maxPrintKeys) {
            log('%d key %s file path %s', k + 1, key, coverage.path);
        }
    });
}
/**
 * Looks at all coverage objects in the given JSON coverage file
 * and if the file is relative, and exists, changes its path to
 * be absolute.
 */
function resolveRelativePaths(nycFilename) {
    const nycCoverage = JSON.parse((0, fs_1.readFileSync)(nycFilename, 'utf8'));
    const coverageKeys = Object.keys(nycCoverage);
    if (!coverageKeys.length) {
        log('⚠️ file %s has no coverage information', nycFilename);
        return;
    }
    log('NYC file %s has %d key(s)', nycFilename, coverageKeys.length);
    let changed = false;
    coverageKeys.forEach((key) => {
        const coverage = nycCoverage[key];
        if (!coverage.path) {
            log('key %s does not have path', key);
            return;
        }
        if (!(0, path_1.isAbsolute)(coverage.path)) {
            if ((0, fs_1.existsSync)(coverage.path)) {
                log('resolving path %s', coverage.path);
                coverage.path = (0, path_1.resolve)(coverage.path);
                changed = true;
            }
            return;
        }
        // path is absolute, let's check if it exists
        if (!(0, fs_1.existsSync)(coverage.path)) {
            log('⚠️ cannot find file %s with hash %s', coverage.path, coverage.hash);
        }
    });
    if (changed) {
        log('resolveRelativePaths saving updated file %s', nycFilename);
        log('there are %d keys in the file', coverageKeys.length);
        (0, fs_1.writeFileSync)(nycFilename, JSON.stringify(nycCoverage, null, 2) + '\n', 'utf8');
    }
}
/**
 * @param filepaths
 * @returns common prefix that corresponds to current folder
 */
function findCommonRoot(filepaths) {
    if (!filepaths.length) {
        log('cannot find common root without any files');
        return;
    }
    // assuming / as file separator
    const splitParts = filepaths.map((name) => name.split('/'));
    const lengths = splitParts.map((arr) => arr.length);
    const shortestLength = Math.min.apply(null, lengths);
    log('shorted file path has %d parts', shortestLength);
    const cwd = process.cwd();
    const commonPrefix = [];
    let foundCurrentFolder;
    for (let k = 0; k < shortestLength; k += 1) {
        const part = splitParts[0][k];
        const prefix = commonPrefix.concat(part).join('/');
        log('testing prefix %o', prefix);
        const allFilesStart = filepaths.every((name) => name.startsWith(prefix));
        if (!allFilesStart) {
            log('stopped at non-common prefix %s', prefix);
            break;
        }
        commonPrefix.push(part);
        const removedPrefixNames = filepaths.map((filepath) => filepath.slice(prefix.length));
        log('removedPrefix %o', removedPrefixNames);
        const foundAllPaths = removedPrefixNames.every((filepath) => (0, fs_1.existsSync)((0, path_1.join)(cwd, filepath)));
        log('all files found at %s? %o', prefix, foundAllPaths);
        if (foundAllPaths) {
            log('found prefix that matches current folder: %s', prefix);
            foundCurrentFolder = prefix;
            break;
        }
    }
    return foundCurrentFolder;
}
function tryFindingLocalFiles(nycFilename) {
    const nycCoverage = JSON.parse((0, fs_1.readFileSync)(nycFilename, 'utf8'));
    const coverageKeys = Object.keys(nycCoverage);
    const filenames = coverageKeys.map((key) => nycCoverage[key].path);
    const commonFolder = findCommonRoot(filenames);
    if (!commonFolder) {
        log('could not find common folder %s', commonFolder);
        return;
    }
    const cwd = process.cwd();
    log('found common folder %s that matches current working directory %s', commonFolder, cwd);
    const length = commonFolder.length;
    let changed = false;
    coverageKeys.forEach((key) => {
        const from = nycCoverage[key].path;
        if (from.startsWith(commonFolder)) {
            const to = (0, path_1.join)(cwd, from.slice(length));
            // ? Do we need to replace the "key" in the coverage object or can we just replace the "path"?
            nycCoverage[key].path = to;
            log('replaced %s -> %s', from, to);
            changed = true;
        }
    });
    if (changed) {
        log('tryFindingLocalFiles saving updated file %s', nycFilename);
        log('there are %d keys in the file', coverageKeys.length);
        (0, fs_1.writeFileSync)(nycFilename, JSON.stringify(nycCoverage, null, 2) + '\n', 'utf8');
    }
}
/**
 * Tries to find source files to be included in the final coverage report
 * using NYC options: extension list, include and exclude.
 */
function findSourceFiles(nycOptions) {
    log('include all files options: %o', {
        all: nycOptions.all,
        include: nycOptions.include,
        exclude: nycOptions.exclude,
        extension: nycOptions.extension
    });
    if (!Array.isArray(nycOptions.extension)) {
        console.error('Expected NYC "extension" option to be a list of file extensions');
        console.error(nycOptions);
        return [];
    }
    const patterns = [];
    if (Array.isArray(nycOptions.include)) {
        patterns.push(...nycOptions.include);
    }
    else if (typeof nycOptions.include === 'string') {
        patterns.push(nycOptions.include);
    }
    else {
        log('using default list of extensions');
        nycOptions.extension.forEach((extension) => {
            patterns.push('**/*' + extension);
        });
    }
    if (Array.isArray(nycOptions.exclude)) {
        const negated = nycOptions.exclude.map((s) => '!' + s);
        patterns.push(...negated);
    }
    else if (typeof nycOptions.exclude === 'string') {
        patterns.push('!' + nycOptions.exclude);
    }
    // always exclude node_modules
    // https://github.com/istanbuljs/nyc#including-files-within-node_modules
    patterns.push('!**/node_modules/**');
    log('searching files to include using patterns %o', patterns);
    const allFiles = (0, tinyglobby_1.globSync)(patterns, { absolute: true });
    return allFiles;
}
/**
 * If the website or unit tests did not load ALL files we need to
 * include, then we should include the missing files ourselves
 * before generating the report.
 *
 * @see https://github.com/cypress-io/code-coverage/issues/207
 */
function includeAllFiles(nycFilename, nycOptions) {
    if (!nycOptions.all) {
        log('NYC "all" option is not set, skipping including all files');
        return;
    }
    const allFiles = findSourceFiles(nycOptions);
    if (log.enabled) {
        log('found %d file(s)', allFiles.length);
        console.error(allFiles.join('\n'));
    }
    if (!allFiles.length) {
        log('no files found, hoping for the best');
        return;
    }
    const nycCoverage = JSON.parse((0, fs_1.readFileSync)(nycFilename, 'utf8'));
    const coverageKeys = Object.keys(nycCoverage);
    const coveredPaths = coverageKeys.map((key) => nycCoverage[key].path.replace(/\\/g, '/'));
    log('coverage has %d record(s)', coveredPaths.length);
    // report on first couple of entries
    if (log.enabled) {
        console.error('coverage has the following first paths');
        console.error(coveredPaths.slice(0, 4).join('\n'));
    }
    let changed = false;
    allFiles.forEach((fullPath) => {
        if (coveredPaths.includes(fullPath)) {
            // all good, this file exists in coverage object
            return;
        }
        log('adding empty coverage for file %s', fullPath);
        changed = true;
        // insert placeholder object for now
        const placeholder = (0, common_utils_1.fileCoveragePlaceholder)(fullPath);
        nycCoverage[fullPath] = placeholder;
    });
    if (changed) {
        log('includeAllFiles saving updated file %s', nycFilename);
        log('there are %d keys in the file', Object.keys(nycCoverage).length);
        (0, fs_1.writeFileSync)(nycFilename, JSON.stringify(nycCoverage, null, 2) + '\n', 'utf8');
    }
}
